src/core_tests/determinism.rs

```diff
@@ -0,0 +1,94 @@
+use crate::{
+    pollers::MockServerGatewayApis,
+    prototype_rust_sdk::{TestRustWorker, WfContext, WorkflowResult},
+    test_help::{
+        build_mock_pollers, canned_histories, mock_core, MockPollCfg, ResponseType,
+        DEFAULT_WORKFLOW_TYPE, TEST_Q,
+    },
+};
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+use temporal_sdk_core_protos::temporal::api::enums::v1::WorkflowTaskFailedCause;
+
+static DID_FAIL: AtomicBool = AtomicBool::new(false);
+pub async fn timer_wf_fails_once(mut ctx: WfContext) -> WorkflowResult<()> {
+    ctx.timer(Duration::from_secs(1)).await;
+    if !DID_FAIL.load(Ordering::Relaxed) {
```

Comment:
Should this be a CAS operation to avoid a race?
I assume it's not critical for this test, just best practice.

Reply:
:+1:  (also makes sense not to use `std::sync::Once` here as it will capture the panic and re-panic everytime)

Reply:
Can't actually race, since WFs aren't called concurrently. Atomics just happen to be a simple way to get a static mutable which is needed for something that will change across invocations